File: atcoder/DP_B.rb
N, K = gets.split.map &:to_i
h = gets.split.map &:to_i
dp = Array.new(N+1,1<<31)
dp[0] = 0
(0..N-2).each{|i|
    (1..K).each{|k|
        if i + k < N
            dp[i+k] = [(h[i+k] - h[i]).abs + dp[i], dp[i+k]].min
        end
    }
}
puts dp[N-1]

File: atcoder/DP_B.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    int N,K,dp[10001],h[10001];
    fill(dp,dp+10001,1e9);
    dp[0] = 0;
    cin >> N >> K;
    for (int i=0;i<N;i++) cin >> h[i];
    for (int i=0;i<N;i++){
        for (int k=1; k<=K;k++){
            if (i + k < N) {
                dp[i + k] = min(abs(h[i+k] - h[i]) + dp[i], dp[i+k]);
            }
        }
    }
    cout << dp[N-1] << endl;
}
